UM2N/generator/unstructured_mesh.py

```diff
@@ -0,0 +1,194 @@
+        # TODO: More detail on Gmsh algorithm number
+        self.mesh_type = mesh_type
+        self._mesh = None
+
+    @property
+    @abc.abstractmethod
+    def corners(self):
+        """
+        :returns: corner vertices of the domain
+        :rtype: tuple
+        """
+        pass
+
+    def generate_mesh(self, res=1e-1, file_path="./temp.msh", remove_file=False):
+        """
+        Generate a mesh at a given resolution level.
+
+        :kwarg res: mesh resolution (element diameter)
+        :type res: float
+        :kwarg file_path: file name for saving the mesh in .msh format
+        :type file_path: str
+        :kwarg remove_file: should the .msh file be removed after generation? (False by
+            default)
+        :type remove_file: bool
+        :returns: mesh generated
+        :rtype: :class:`firedrake.mesh.MeshGeometry`
+        """
+        gmsh.initialize()
+        gmsh.model.add("t1")
+        self.lc = res
+        self._points = [
+            gmsh.model.geo.addPoint(*corner, 0, self.lc) for corner in self.corners
+        ]
+        self._lines = [
+            gmsh.model.geo.addLine(point, point_next)
+            for point, point_next in zip(
+                self._points, self._points[1:] + [self._points[0]]
+            )
+        ]
+        gmsh.model.geo.addCurveLoop([i + 1 for i in range(len(self._points))], 1)
+        gmsh.model.geo.addPlaneSurface([1], 1)
+        gmsh.model.geo.synchronize()
+        gmsh.option.setNumber("Mesh.Algorithm", self.mesh_type)
+        for i, line_tag in enumerate(self._lines):
+            gmsh.model.addPhysicalGroup(1, [line_tag], i + 1)
+            gmsh.model.setPhysicalName(1, i + 1, "Boundary " + str(i + 1))
+        gmsh.model.addPhysicalGroup(2, [1], name="My surface")
+        gmsh.model.mesh.generate(2)
+        gmsh.write(file_path)
+        gmsh.finalize()
+        self.num_boundary = len(self._lines)
+        self._mesh = Mesh(file_path)
+        if remove_file:
+            os.remove(file_path)
+        return self._mesh
+
+    def load_mesh(self, file_path):
+        """
+        Load a mesh from a file saved in .msh format.
+
+        :arg file_path: filename including the .msh extension
+        :type file_path: str
+        :returns: mesh loaded from file
+        :rtype: :class:`firedrake.mesh.MeshGeometry`
+        """
+        self._mesh = Mesh(file_path)
+        return self._mesh
+
+
+class UnstructuredSquareMeshGenerator(UnstructuredMeshGenerator):
+    """
+    Generate an unstructured mesh of a 2D square domain using Gmsh.
+    """
+
+    @property
+    def corners(self):
+        """
+        :returns: corner vertices of the square domain
+        :rtype: tuple
+        """
+        return ((0, 0), (self.scale, 0), (self.scale, self.scale), (0, self.scale))
+
+
+class UnstructuredRandomPolygonalMeshGenerator(UnstructuredMeshGenerator):
+    """
+    Create a random polygonal mesh by spliting the edge of a
+    square randomly.
+    """
```

Comment:
Good idea, will do that and extend the tests.

Reply:
Addressed in 18561f8. I think it needs to be on the `generate_mesh` method in order to get reproducible mesh generation. I decided to delegate expanding the test suite for later - see #51.

Reply:
This doesn't work unfortunately. Two issues:

1. You are using both `random` and `np.random` so both `random.seed` and `np.random.seed` need to be set. Can we avoid using one of them?
2. The randomness happens in the `corners` attribute which is only computed once and stored in `._corners`. So if we do `mesh1 = poly_mesh_generator.generate_mesh(seed=0)` and `mesh2 = poly_mesh_generator.generate_mesh(seed=1)`, then `mesh1` and `mesh2` are identical since their `_corners` property is the same. So I guess the simplest solution is to pass seed `__init__`?

But more generally... connected to point 2, I find the name `*Generator` a bit misleading since I'd then expect it to generate a different mesh every time we call `generate_mesh()`. Why not just call it `UnstructuredSquareMesh` (and similarly for others) since `generate_mesh()` always generates the same mesh? But not a big deal, just thought I'd mention it as part of point 2 above :)

Reply:
Btw, I was curious about best practices with seeding the generator. The general consensus is that we should avoid using `np.random.seed()` because this is a global Random Number Generator (RNG). So by setting `np.random.seed()` in `UnstructuredRandomPolygonalMeshGenerator`, we are affecting the rest of the user code as well. See e.g. https://albertcthomas.github.io/good-practices-random-number-generators/ . We should instead do
```
rng = np.random.default_rng(seed)
rng.uniform(...)
```

Reply:
In the interests of getting this PR merged, I reverted the seeding in b9f2dcf3ccffedbf7578ec578aa6ad259120953b.

I've opened issues #52 and #53. With the naming I want to be careful to make sure any `Mesh` objects are consistent with Firedrake ones so it will require some thought.